src/OpenTelemetry.Instrumentation.AspNet.Win/Implementation/HttpInListener.cs

```diff
@@ -154,15 +172,36 @@ public override void OnStopActivity(Activity activity, object payload)
                 else if (routeData.Route is Route route)
                 {
                     // MVC + WebAPI traditional routing & MVC attribute routing flow here.
-
                     template = route.Url;
                 }
 
                 if (!string.IsNullOrEmpty(template))
                 {
                     // Override the name that was previously set to the path part of URL.
-                    activity.DisplayName = template;
-                    activity.AddTag(SpanAttributeConstants.HttpRouteKey, template);
+                    activityToEnrich.DisplayName = template;
+                    activityToEnrich.AddTag(SpanAttributeConstants.HttpRouteKey, template);
+                }
+            }
+
+            if (!(this.options.TextFormat is TraceContextFormatActivity))
+            {
+                if (activity.OperationName.Equals(ActivityNameByHttpInListener))
+                {
+                    // If instrumentation started a new Activity, it must
+                    // be stopped here.
+                    activity.Stop();
+
+                    // Restore the original activity as Current.
+                    var activityByAspNet = (Activity)activity.GetCustomProperty("ActivityByAspNet");
+                    Activity.Current = activityByAspNet;
+                }
+                else if (activity.OperationName.Equals("Microsoft.AspNet.HttpReqIn.Start"))
+                {
+                    // This block is hit if Asp.Net did restore Current to its own activity,
+                    // then we need to retrieve the one created by HttpInListener
+                    // and stop it.
+                    var activityByHttpInListener = (Activity)activity.GetCustomProperty("ActivityByHttpInListener");
+                    activityByHttpInListener.Stop();
```

Comment:
One last thing :) When we stop "newOne" here do we need to restore Activity.Current back to libraryActivity (since the Stop will call `Activity.SetCurrent(Parent)`)? Something like:
```csharp
                else if (activity.OperationName.Equals("Microsoft.AspNet.HttpReqIn.Start"))
                {
                    // This block is hit if Asp.Net did restore Current to its own activity,
                    // then we need to retrieve the one created by HttpInListener
                    // and stop it.
                    var activityByHttpInListener = (Activity)activity.GetCustomProperty("ActivityByHttpInListener");
                    activityByHttpInListener.Stop();
                    Activity.Current = activity; // Put back libraryActivity
                }
```

Reply:
Right. Added tests too for this.